## Practicing pandas 

In [2]:
import pandas as pd

In [3]:
file_path = "C:\MyDocs\Machine learning\Chicago_Crime_Data-v2.csv"

In [4]:
df = pd.read_csv(file_path)

In [5]:
df.head()

ID CASE_NUMBER                    DATE                     BLOCK  \
0   3512276    HK587712  08/28/2004 05:50:56 PM        047XX S KEDZIE AVE   
1   3406613    HK456306  06/26/2004 12:40:00 PM  009XX N CENTRAL PARK AVE   
2   8002131    HT233595  04/04/2011 05:45:00 AM        043XX S WABASH AVE   
3   7903289    HT133522  12/30/2010 04:30:00 PM      083XX S KINGSTON AVE   
4  10402076    HZ138551  02/02/2016 07:30:00 PM           033XX W 66TH ST   

  IUCR PRIMARY_TYPE                    DESCRIPTION  \
0  890        THEFT                  FROM BUILDING   
1  820        THEFT                 $500 AND UNDER   
2  820        THEFT                 $500 AND UNDER   
3  840        THEFT  FINANCIAL ID THEFT: OVER $300   
4  820        THEFT                 $500 AND UNDER   

           LOCATION_DESCRIPTION  ARREST  DOMESTIC  \
0            SMALL RETAIL STORE   False     False   
1                         OTHER   False     False   
2  NURSING HOME/RETIREMENT HOME   False     False   
3                     RESIDENCE   False     False   
4                         ALLEY   False     False   

               ...                WARD  COMMUNITY_AREA_NUMBER  FBICODE  \
0              ...                14.0                   58.0        6   
1              ...                27.0                   23.0        6   
2              ...                 3.0                   38.0        6   
3              ...                 7.0                   46.0        6   
4              ...                15.0                   66.0        6   

   X_COORDINATE Y_COORDINATE  YEAR               UPDATEDON   LATITUDE  \
0     1155838.0    1873050.0  2004  02/10/2018 03:50:01 PM  41.807441   
1     1152206.0    1906127.0  2004  02/28/2018 03:56:25 PM  41.898280   
2     1177436.0    1876313.0  2011  02/10/2018 03:50:01 PM  41.815933   
3     1194622.0    1850125.0  2010  02/10/2018 03:50:01 PM  41.743665   
4     1155240.0    1860661.0  2016  02/10/2018 03:50:01 PM  41.773455   

   LONGITUDE                       LOCATION  
0 -87.703956    (41.8074405, -87.703955849)  
1 -87.716406  (41.898279962, -87.716405505)  
2 -87.624642  (41.815933131, -87.624642127)  
3 -87.562463  (41.743665322, -87.562462756)  
4 -87.706480  (41.773455295, -87.706480471)  

[5 rows x 22 columns]

### Get all column names

In [6]:
df.columns

Index(['ID', 'CASE_NUMBER', 'DATE', 'BLOCK', 'IUCR', 'PRIMARY_TYPE',
       'DESCRIPTION', 'LOCATION_DESCRIPTION', 'ARREST', 'DOMESTIC', 'BEAT',
       'DISTRICT', 'WARD', 'COMMUNITY_AREA_NUMBER', 'FBICODE', 'X_COORDINATE',
       'Y_COORDINATE', 'YEAR', 'UPDATEDON', 'LATITUDE', 'LONGITUDE',
       'LOCATION'],
      dtype='object')

In [7]:
list(df.columns)

['ID',
 'CASE_NUMBER',
 'DATE',
 'BLOCK',
 'IUCR',
 'PRIMARY_TYPE',
 'DESCRIPTION',
 'LOCATION_DESCRIPTION',
 'ARREST',
 'DOMESTIC',
 'BEAT',
 'DISTRICT',
 'WARD',
 'COMMUNITY_AREA_NUMBER',
 'FBICODE',
 'X_COORDINATE',
 'Y_COORDINATE',
 'YEAR',
 'UPDATEDON',
 'LATITUDE',
 'LONGITUDE',
 'LOCATION']

## Find out dimensions of the data frame

In [8]:
df.shape

(533, 22)

## Explore DataFrame 

In [9]:
df.describe()

ID         BEAT    DISTRICT        WARD  \
count  5.330000e+02   533.000000  533.000000  490.000000   
mean   6.228301e+06  1187.037523   11.195122   22.608163   
std    3.027454e+06   660.592263    6.496631   13.080119   
min    2.114900e+04   111.000000    1.000000    1.000000   
25%    3.420516e+06   711.000000    6.000000   12.000000   
50%    6.231712e+06  1111.000000   10.000000   24.000000   
75%    8.875058e+06  1634.000000   16.000000   32.000000   
max    1.127717e+07  2535.000000   25.000000   50.000000   

       COMMUNITY_AREA_NUMBER  X_COORDINATE  Y_COORDINATE        YEAR  \
count             490.000000  5.290000e+02  5.290000e+02   533.00000   
mean               37.542857  1.162876e+06  1.886223e+06  2008.34334   
std                21.409514  1.650197e+04  3.009980e+04     4.94940   
min                 1.000000  1.100658e+06  1.814512e+06  2001.00000   
25%                23.000000  1.151715e+06  1.860454e+06  2004.00000   
50%                30.000000  1.163235e+06  1.891447e+06  2008.00000   
75%                58.000000  1.174618e+06  1.907711e+06  2012.00000   
max                77.000000  1.204126e+06  1.951001e+06  2018.00000   

         LATITUDE   LONGITUDE  
count  529.000000  529.000000  
mean    41.843429  -87.677818  
std      0.082798    0.060073  
min     41.645796  -87.905227  
25%     41.772556  -87.718166  
50%     41.857737  -87.676733  
75%     41.902298  -87.634677  
max     42.021178  -87.528223

### It seems although there are 533 rows not all of them have values e.g. COMMUNITY_AREA_NUMBER has 490 values. Let's find out is it really the case

In [10]:
df.value_counts()
#did not work

AttributeError: 'DataFrame' object has no attribute 'value_counts'

In [11]:
df['COMMUNITY_AREA_NUMBER'].value_counts()

25.0    43
23.0    22
68.0    21
29.0    16
28.0    16
8.0     15
71.0    14
24.0    13
67.0    12
66.0    12
6.0     11
49.0    11
69.0    11
61.0    10
58.0    10
26.0    10
43.0     9
31.0     9
22.0     9
46.0     9
44.0     8
19.0     8
27.0     8
16.0     8
42.0     8
38.0     8
70.0     8
30.0     8
63.0     7
32.0     7
        ..
76.0     4
39.0     4
5.0      4
62.0     4
3.0      4
51.0     4
45.0     4
73.0     3
10.0     3
17.0     3
65.0     3
20.0     3
50.0     3
4.0      3
64.0     2
77.0     2
37.0     2
54.0     2
56.0     2
41.0     2
52.0     2
11.0     2
18.0     2
59.0     2
33.0     1
60.0     1
55.0     1
47.0     1
75.0     1
12.0     1
Name: COMMUNITY_AREA_NUMBER, Length: 70, dtype: int64

In [13]:
df['COMMUNITY_AREA_NUMBER'].isnull().value_counts()

False    490
True      43
Name: COMMUNITY_AREA_NUMBER, dtype: int64

### Let's find out similar data for all columns

In [14]:
df.isnull()

ID  CASE_NUMBER   DATE  BLOCK   IUCR  PRIMARY_TYPE  DESCRIPTION  \
0    False        False  False  False  False         False        False   
1    False        False  False  False  False         False        False   
2    False        False  False  False  False         False        False   
3    False        False  False  False  False         False        False   
4    False        False  False  False  False         False        False   
5    False        False  False  False  False         False        False   
6    False        False  False  False  False         False        False   
7    False        False  False  False  False         False        False   
8    False        False  False  False  False         False        False   
9    False        False  False  False  False         False        False   
10   False        False  False  False  False         False        False   
11   False        False  False  False  False         False        False   
12   False        False  False  False  False         False        False   
13   False        False  False  False  False         False        False   
14   False        False  False  False  False         False        False   
15   False        False  False  False  False         False        False   
16   False        False  False  False  False         False        False   
17   False        False  False  False  False         False        False   
18   False        False  False  False  False         False        False   
19   False        False  False  False  False         False        False   
20   False        False  False  False  False         False        False   
21   False        False  False  False  False         False        False   
22   False        False  False  False  False         False        False   
23   False        False  False  False  False         False        False   
24   False        False  False  False  False         False        False   
25   False        False  False  False  False         False        False   
26   False        False  False  False  False         False        False   
27   False        False  False  False  False         False        False   
28   False        False  False  False  False         False        False   
29   False        False  False  False  False         False        False   
..     ...          ...    ...    ...    ...           ...          ...   
503  False        False  False  False  False         False        False   
504  False        False  False  False  False         False        False   
505  False        False  False  False  False         False        False   
506  False        False  False  False  False         False        False   
507  False        False  False  False  False         False        False   
508  False        False  False  False  False         False        False   
509  False        False  False  False  False         False        False   
510  False        False  False  False  False         False        False   
511  False        False  False  False  False         False        False   
512  False        False  False  False  False         False        False   
513  False        False  False  False  False         False        False   
514  False        False  False  False  False         False        False   
515  False        False  False  False  False         False        False   
516  False        False  False  False  False         False        False   
517  False        False  False  False  False         False        False   
518  False        False  False  False  False         False        False   
519  False        False  False  False  False         False        False   
520  False        False  False  False  False         False        False   
521  False        False  False  False  False         False        False   
522  False        False  False  False  False         False        False   
523  False        False  False  False  False         False        False   
524  False        False  False  F

In [15]:
df.isnull().value_counts()

AttributeError: 'DataFrame' object has no attribute 'value_counts'

In [16]:
type(df['COMMUNITY_AREA_NUMBER'])

pandas.core.series.Series

In [17]:
'value_counts' in dir(df['COMMUNITY_AREA_NUMBER'])

True

## 'value_counts' method is available only on pandas series

In [19]:
for each_col in df.columns:
    print(df[each_col].isnull().value_counts())

False    533
Name: ID, dtype: int64
False    533
Name: CASE_NUMBER, dtype: int64
False    533
Name: DATE, dtype: int64
False    533
Name: BLOCK, dtype: int64
False    533
Name: IUCR, dtype: int64
False    533
Name: PRIMARY_TYPE, dtype: int64
False    533
Name: DESCRIPTION, dtype: int64
False    533
Name: LOCATION_DESCRIPTION, dtype: int64
False    533
Name: ARREST, dtype: int64
False    533
Name: DOMESTIC, dtype: int64
False    533
Name: BEAT, dtype: int64
False    533
Name: DISTRICT, dtype: int64
False    490
True      43
Name: WARD, dtype: int64
False    490
True      43
Name: COMMUNITY_AREA_NUMBER, dtype: int64
False    533
Name: FBICODE, dtype: int64
False    529
True       4
Name: X_COORDINATE, dtype: int64
False    529
True       4
Name: Y_COORDINATE, dtype: int64
False    533
Name: YEAR, dtype: int64
False    533
Name: UPDATEDON, dtype: int64
False    529
True       4
Name: LATITUDE, dtype: int64
False    529
True       4
Name: LONGITUDE, dtype: int64
False    529
True       4
N

In [24]:
for each_col in df.columns:
    stat = df[each_col].isnull().value_counts()
    print(stat)
    print(type(stat))
    print(stat.index)
    break

False    533
Name: ID, dtype: int64
<class 'pandas.core.series.Series'>
Index([False], dtype='object')


In [30]:
#resetting index
for each_col in df.columns:
    stat = df[each_col].isnull().value_counts()
    #print(stat)
    #print(type(stat))
    print(stat.reset_index())
    print(stat.index)
    break

   index   ID
0  False  533
Index([False], dtype='object')


In [37]:
#print only columns having nulls and their counts

for each_col in df.columns:
    stat = df[each_col].isnull().value_counts()
    if any(stat.index):
        print("{col_name} has {cnt} null values".format(col_name=each_col, cnt=stat[True]))


WARD has 43 null values
COMMUNITY_AREA_NUMBER has 43 null values
X_COORDINATE has 4 null values
Y_COORDINATE has 4 null values
LATITUDE has 4 null values
LONGITUDE has 4 null values
LOCATION has 4 null values


## Now let's drop null values

In [41]:
newdf = df.dropna()

In [47]:
def get_columns_with_null_values(df):
    for each_col in df.columns:
        stat = df[each_col].isnull().value_counts()
        if any(stat.index):
            print("{col_name} has {cnt} null values".format(col_name=each_col, cnt=stat[True]))

#get_columns_with_null_values(newdf)
print(type(df.isnull()))

        

<class 'pandas.core.frame.DataFrame'>


In [48]:
df.dropna().shape

(487, 22)

In [50]:
df.dropna(inplace=True)

In [51]:
df.shape

(487, 22)

In [57]:
any(df.isnull())

True

## Does this mean there are null values present? Let's find out

In [62]:
get_columns_with_null_values(df)
#No output

In [64]:
for each in df.isnull():
    print(each)
    print(any(each))
    break

ID
True


In [67]:
df.isnull().__iter__

<bound method NDFrame.__iter__ of         ID  CASE_NUMBER   DATE  BLOCK   IUCR  PRIMARY_TYPE  DESCRIPTION  \
0    False        False  False  False  False         False        False   
1    False        False  False  False  False         False        False   
2    False        False  False  False  False         False        False   
3    False        False  False  False  False         False        False   
4    False        False  False  False  False         False        False   
5    False        False  False  False  False         False        False   
6    False        False  False  False  False         False        False   
7    False        False  False  False  False         False        False   
8    False        False  False  False  False         False        False   
9    False        False  False  False  False         False        False   
10   False        False  False  False  False         False        False   
11   False        False  False  False  False         False        

In [76]:
for each in df.isnull():
    print(each, any(df.isnull()[each]))
    #print(df.isnull()[each])

ID False
CASE_NUMBER False
DATE False
BLOCK False
IUCR False
PRIMARY_TYPE False
DESCRIPTION False
LOCATION_DESCRIPTION False
ARREST False
DOMESTIC False
BEAT False
DISTRICT False
WARD False
COMMUNITY_AREA_NUMBER False
FBICODE False
X_COORDINATE False
Y_COORDINATE False
YEAR False
UPDATEDON False
LATITUDE False
LONGITUDE False
LOCATION False


In [78]:
for each in df:
    print(each, type(each))
    break

ID <class 'str'>


## Before using 'any' or 'all' make sure what is the iterator returning. 
## For DataFrame the iterator returns column names, so dont directly use 'any' or 'all' on it.